Add clang CI test

PointKernel opened the pull request:
Fixes #744
Fixes #128

This PR adds a CI task to test building with Clang as the host compiler and addresses associated build warnings and errors.

PointKernel committed:
Add clang build matrix

PointKernel committed:
Update

PointKernel committed:
Fix Clang build warnings

PointKernel committed:
Fix an example warning

PointKernel committed:
Disable benchmark build for clang CI

PointKernel committed:
Update copyright years

PointKernel commented:
/ok to test f27cfc9

PointKernel committed:
Update GPU archs

sleeepyjack commented:
/ok to test [ad63e43](https://github.com/NVIDIA/cuCollections/pull/743/commits/ad63e43255225d2aec7fcbc5901273d380b03b16)

PointKernel committed:
Remove clang test CI

PointKernel committed:
Update

PointKernel committed:
Updates

PointKernel committed:
Fix CI success job: add doxygen-check dependency and restore skipped condition

PointKernel committed:
Fix CI issues: remove skipped condition from ci-success and add missing devcontainer configs

- Remove skipped condition from ci-success job to allow intentionally skipped test jobs
- Add missing devcontainer configs for cuda12.0-gcc11 and cuda12.8-llvm16 to match matrix.yml

PointKernel committed:
Enhance Clang devcontainer following CCCL patterns

- Add build volume management for better performance
- Include additional VSCode extensions (clang-format, nsight, cmake-tools)
- Enhanced editor settings with format-on-save and better clangd config
- Make GPU requirement optional for better compatibility
- Add build infix environment variable for consistency

PointKernel committed:
Fix devcontainer files formatting

- Add missing newline to cuda12.0-gcc11/devcontainer.json
- Standardize cuda12.8-llvm16/devcontainer.json to match base template
- Remove custom enhancements to follow project standards
- Generated using .devcontainer/make_devcontainers.sh

PointKernel committed:
Add CI debug output to identify failing jobs

PointKernel committed:
Enhance CI debug to show matrix job details

PointKernel committed:
Add compute-matrix debugging to identify empty strategy

PointKernel committed:
Fix key generator init

PointKernel committed:
Remove unused devcontainer

PointKernel committed:
Revert debugging info

PointKernel committed:
Cleanups

PointKernel committed:
Rework

PointKernel committed:
Treat warnings as errors

PointKernel committed:
Debug

PointKernel committed:
Update copyright years

sleeepyjack commented on .github/workflows/dispatch-build-and-test.yml:
```diff
@@ -1,4 +1,4 @@
-# SPDX-FileCopyrightText: Copyright (c) 2023 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
no changes in this file so technically we could leave the copyright year as is.

sleeepyjack commented on include/cuco/detail/hash_functions/murmurhash3.cuh:
```diff
@@ -233,6 +233,8 @@ struct MurmurHash3_32 {
         k1 = rotl32(k1, 15);
         k1 *= c2;
         h1 ^= k1;
+        [[fallthrough]];
+      case 0: break;
```
This should also go into the CCCL port

sleeepyjack commented on include/cuco/detail/utility/cuda.cuh:
```diff
@@ -48,7 +48,7 @@ __device__ constexpr int32_t warp_size() noexcept { return 32; }
  * @return The global thread index
  */
-__device__ static index_type global_thread_id() noexcept
+[[maybe_unused]] __device__ static index_type global_thread_id() noexcept
```
Why is this needed?

PointKernel replied:
This is related to an odd compilation warning in Clang: if a function is declared `static` but isn't used within the translation unit, Clang emits an "unused function" warning. The most reliable way I've found to make the code compile cleanly with both GCC and Clang is to add the `[[maybe_unused]]` attribute.

sleeepyjack commented on .github/actions/compute-matrix/compute-matrix.py:
```diff
@@ -0,0 +1,166 @@
+#!/usr/bin/env python3
```
What's the difference between the bash and the python implementation?

PointKernel replied:
Removed as it's only for testing

sleeepyjack reviewed (commented):
It seems GH actions is encountering a syntax error: https://github.com/NVIDIA/cuCollections/actions/runs/16486736912/workflow

PointKernel committed:
Revert changes

PointKernel committed:
testing

PointKernel committed:
Add clang CI back

PointKernel committed:
Update matrix to avoid empty entry

PointKernel committed:
Remove debug output

PointKernel committed:
Updat oldest llvm version

PointKernel committed:
Add special handling for 12.0 with clang

PointKernel commented:
/ok to test 3e6e64c

PointKernel committed:
Updates

PointKernel committed:
Updates

PointKernel committed:
Update CI matrix infp

PointKernel commented on include/cuco/detail/utility/cuda.cuh:
```diff
@@ -41,14 +41,14 @@ namespace cuco {
  * @return The global thread index
  */
-__device__ static index_type global_thread_id() noexcept
+[[nodiscard]] __device__ inline index_type global_thread_id() noexcept
```
`inline` is needed to silence the multiple definition error with clang

PointKernel committed:
Updates

PointKernel committed:
Updates

PointKernel committed:
Clean up CI matrix

PointKernel committed:
Clean up devcontainer

PointKernel committed:
Minor updates

PointKernel committed:
Update os version for the llvm14 container

PointKernel committed:
Updates

sleeepyjack reviewed (approved):
Thanks for going through this CI struggle. Nice work!
